pkg/util/watch/until.go

```diff
@@ -0,0 +1,403 @@
+	return fw.Client.Watch(fw.Context, objList, optsList...)
+}
+
+// listOptions builds a set of List filters for the flexWatcher's Key and Labels.
+func (fw *flexWatcher) listOptions() *client.ListOptions {
+	opts := client.ListOptions{}
+	if len(fw.Key.Namespace) > 0 {
+		opts.Namespace = fw.Key.Namespace
+	}
+	if len(fw.Key.Name) > 0 {
+		opts.FieldSelector = fields.OneTermEqualSelector(metav1.ObjectNameField, fw.Key.Name)
+	}
+	if len(fw.Labels) > 0 {
+		opts.LabelSelector = client.MatchingLabelsSelector{
+			Selector: labels.SelectorFromSet(fw.Labels),
+		}
+	}
+	return &opts
+}
+
+// MergeListOptions merges two sets of ListOptions.
+// - For Namespace, Limit, and Continue, at most one may be specified, unless they're the same
+// - For FieldSelector and LabelSelector, the result requires both filters to match (AND)
+// - For Raw, at most one may be specified
+func MergeListOptions(a, b *client.ListOptions) (*client.ListOptions, error) {
+	switch {
+	case a == nil && b == nil:
+		return nil, nil
+	case a == nil:
+		return b, nil
+	case b == nil:
+		return a, nil
+	}
+
+	c := client.ListOptions{}
+
+	switch {
+	case a.Namespace != "" && b.Namespace != "":
+		if a.Namespace != b.Namespace {
+			return nil, errors.Errorf("cannot merge two different namespaces: %s & %s",
+				a.Namespace, b.Namespace)
+		}
+		c.Namespace = a.Namespace
+	case a.Namespace != "":
+		c.Namespace = a.Namespace
+	case b.Namespace != "":
+		c.Namespace = b.Namespace
+	}
+
+	switch {
+	case a.FieldSelector != nil && b.FieldSelector != nil:
+		c.FieldSelector = fields.AndSelectors(a.FieldSelector, b.FieldSelector)
+	case a.FieldSelector != nil:
+		c.FieldSelector = a.FieldSelector
+	case b.FieldSelector != nil:
+		c.FieldSelector = b.FieldSelector
+	}
+
+	switch {
+	case a.LabelSelector != nil && b.LabelSelector != nil:
+		bReqs, selectable := b.LabelSelector.Requirements()
+		if selectable {
+			c.LabelSelector = a.LabelSelector.Add(bReqs...)
+		} else {
+			c.LabelSelector = a.LabelSelector
+		}
+	case a.FieldSelector != nil:
+		c.LabelSelector = a.LabelSelector
+	case b.FieldSelector != nil:
+		c.LabelSelector = b.LabelSelector
+	}
+
+	switch {
+	case a.Limit > 0 && b.Limit > 0:
+		if a.Limit != b.Limit {
+			return nil, errors.Errorf("cannot merge two different limits: %d & %d",
+				a.Limit, b.Limit)
+		}
+		c.Limit = a.Limit
+	case a.Limit > 0:
+		c.Limit = a.Limit
+	case b.Limit > 0:
+		c.Limit = b.Limit
+	}
+
+	switch {
+	case a.Continue != "" && b.Continue != "":
+		if a.Continue != b.Continue {
+			return nil, errors.Errorf("cannot merge two different continue tokens: %s & %s",
+				a.Continue, b.Continue)
+		}
+		c.Continue = a.Continue
+	case a.Continue != "":
+		c.Continue = a.Continue
+	case b.Continue != "":
+		c.Continue = b.Continue
+	}
+
+	switch {
+	case a.Raw != nil && b.Raw != nil:
+		// TODO: It should be possible to merge raw ListOptions, but we don't need it yet
```

Comment:
Curious why we don't need it - do we just not define any raw `ListOptions` in code currently?

Reply:
The only code using this conversion method today is the flexWatcher, implementing ListerWatcher for UntilDeletedWithSync. 

UntilDeletedWithSync calls client-go's UntilWithSync which builds an Informer which wraps a Reflector. So the Reflector is the only thing calling the List & Watch methods on the flexWatcher. It specifies ResourceVersion, TimeoutSeconds, and AllowWatchBookmarks for Watch (which adds Watch=true) and ResourceVersion, ResourceVersionMatch, Limit, and Continue for List.

The metav1.ListOptions then passes through ConvertListOptions, MergeListOptions, and UnrollListOptions in the flexWatcher List & Watch methods. 
- ConvertListOptions converts the metav1.ListOptions from the Reflector into client.ListOptions (including some Raw options that don't have client.ListOptions fields)
- MergeListOptions merges the options from the Reflector with the options from flexWatcher.listOptions()
- UnrollListOptions converts the client.ListOptions into an array of client.ListOption

So there's no need to merge the Raw options unless we start using these conversion functions elsewhere, which we might in the future, since they're kinda handy when you have a client.Client and want to make an Informer to watch things.

Reply:
Ack